paddle/framework/framework.md

```diff
@@ -0,0 +1,205 @@
+
+The global and nested local scopes form a hierarchy.  The following Python functions make it convenient to program scopes:
+
+1. `paddle.scope.current()` returns the current scope, which defaults to
+1. `paddle.scope.global()`, which returns the top-level scope.
+
+C++ code shouldn't maintain global status like the current scope to prevent unexpected inconsistancy.
+
+```cpp
+class Scope {
+ public:
+  Scope() : parent_(nullptr) {} // Constructor creates only global scopes. 
+  
+  Variable* FindVar(std::string name); // Find in the hierrachy or return null.
+  Variable* CreateVar(StringPiece name, Variable*); // Find or create.
+  
+  Scope* CreateScope(StringPiece name); // Finds or creates a sub-scope.
+  void DeleteScope(Scope*);             // Delete a sub-scope or raise exception
+  
+ private:
+  std::map<std::string /*name*/, std::unique_ptr<Variable> > vars_;
+  std::shared_ptr<Scope> parent_;
+  std::vector<Scope*> children_;
+  
+  Mutex mutex_;  // Make this class thread-safe.
+};
+```
+
+## Execution and Context
+
+A neural network is a program.  Training or inference is to execute it.  The runtime environment of execution is known as a *context*:
+
+1. a scope,
+1. device(s), or places,
+1. a flag indicates if we are training and should we create gradient operations and run backward
+
+and can be defined as
+
+```cpp
+struct Context {
+  Scope* scope_;
+  std::vector<Place> places_; // a network might run on multiple devices.
+  bool training_;
+};
+```
+
+The Python API `paddle.train` can prepare a context before it calls C++ code `Operator::Run(const Context&)`.
+
+As an example, `paddle::operator::Recurrent::Run(const Context& ctx)` can then create a new scope by calling `ctx.CreateScope`, and run the step-net with a new context around the new scope:
+
+```cpp
+class Recurrent {
+ public:
+  void Run(const Context& ctx) {
+    auto subscope = ctx.scope_.CreateScope("");
+    step_net_.Run(Context{subscope, ctx.places_, ctx.training_});
+    if (!ctx.training_) {
+      ctx.scope_.DeleteScope(subscope);
+    }
+};
+```
+
+Another example is that the Gemm operator need to create a tensor on `Context::places_[0]` and assign the tensor to its output variable:
+
+```cpp
+class Gemm {
+ public:
+  void Run(const Context& ctx) {
+    if (paddle::platform::IsGPUPlace(ctx.places_[0])) {
+      cuDNNGemm(
+         Output(0).mutable_data<float>(ctx.places_[0], DerivedSizeFromInputs()),
+         ...);
```

Comment:
cudnnGemm needs to acquire a cudnnHandle to finish computation.
And nearly all the computation in Run method need to acquire a eigen device.
Please refer to #2648